src/daemon/auth/providers/config_file.rs

```diff
@@ -0,0 +1,386 @@
+
+                Ok(Some(self.auth_from_session(&session)?))
+            }
+            _ => Ok(None),
+        };
+
+        if log_enabled!(log::Level::Trace) {
+            trace!("Authentication result: {:?}", res);
+        }
+
+        res
+    }
+
+    pub fn get_login_url(&self) -> KrillResult<HttpResponse> {
+        // Direct Lagosta to show the user the Lagosta API token login form
+        Ok(HttpResponse::text_no_cache(UI_LOGIN_ROUTE_PATH.into()))
+    }
+
+    pub fn login(&self, request: &HyperRequest) -> KrillResult<LoggedInUser> {
+        use scrypt::scrypt;
+
+        let auth = match self.get_auth(request) {
+            Some(auth) => auth,
+            None => {
+                trace!("Missing or incomplete credentials for login attempt");
+                return Err(Error::ApiInvalidCredentials(
+                    "Missing credentials".to_string(),
+                ))
+            }
+        };
+
+        // Do NOT bail out if the user is not known because then the
+        // unknown user path would return very quickly
+        // compared to the known user path and timing differences can aid
+        // attackers.
+        let (user_password_hash, user_salt) =
+            match self.users.get(&auth.username) {
+                Some(user) => {
+                    (user.password_hash.as_ref(), user.salt.as_ref())
+                }
+                None => (FAKE_PASSWORD_HASH, FAKE_SALT),
+            };
+
+        let username = auth.username.trim().nfkc().collect::<String>();
+        let password = auth.password.trim().nfkc().collect::<String>();
+
+        // hash twice with two different salts
+        // legacy hashing strategy to be compatible with lagosta
+        let params = scrypt::Params::new(
+            PW_HASH_LOG_N,
+            PW_HASH_R,
+            PW_HASH_P,
+            scrypt::Params::RECOMMENDED_LEN,
+        )
+        .unwrap();
+        let weak_salt = format!("krill-lagosta-{username}");
+        let weak_salt = weak_salt.nfkc().collect::<String>();
+
+        let mut interim_hash: [u8; 32] = [0; 32];
+        scrypt(
+            password.as_bytes(),
+            weak_salt.as_bytes(),
+            &params,
+            &mut interim_hash,
+        )
+        .unwrap();
+
+        let strong_salt: Vec<u8> = hex::decode(user_salt).unwrap();
+        let mut hashed_hash: [u8; 32] = [0; 32];
+        scrypt(
+            &interim_hash,
+            strong_salt.as_slice(),
+            &params,
+            &mut hashed_hash,
+        )
+        .unwrap();
+
+        let encoded_hash = hex::encode(hashed_hash);
+
+        // And now finally check the user, so that both known and
+        // unknown user code paths do the same work
+        // and don't result in an obvious timing difference between
+        // the two scenarios which could potentially
+        // be used to discover user names.
+        if encoded_hash != user_password_hash {
```

Comment:
I know this was not changed recently, but I don't like the "succeed unless it fails" flow for authentication. I prefer an explicit "fail unless it succeeds" flow, as if authentication becomes more complex in the future, then making a small mistake where the checks are all bypassed becomes easier.

Reply:
I agree with your point, but in this case the code would become super unreadable. It would drift right badly and all the errors would have to be created in else branches far away from the check.